src/build_loop.rs

```diff
@@ -9,32 +9,157 @@ use crate::project::roots;
 use crate::project::roots::Roots;
 use crate::project::Project;
 use crate::watch::{DebugMessage, RawEventError, Reason, Watch};
+use crate::NixFile;
+use serde::{
+    de::{self, Visitor},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
 use std::path::PathBuf;
 use std::sync::mpsc::{channel, Sender};
 
 /// Builder events sent back over `BuildLoop.tx`.
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
 pub enum Event {
-    /// The build has started
-    Started(Reason),
-    /// The build completed successfully
-    Completed(BuildResults),
-    /// The build command returned a failing exit status
-    Failure(BuildExitFailure),
+    /// A heartbeat to send over sockets to keep them alive
+    Heartbeat,
```

Comment:
Design question: Do we absolutely *need* to send a heartbeat? I’d like to keep the state machine as simple as possible. How about we just try to send a message and throw the client out if we can’t send it?

In that case, we would just stop the client command if the connection is broken for whatever reason (let’s assume the socket can connect always if the daemon is running). No fallbacks, no retries, no heartbeats.

Reply:
The client (all socket::Readers iiuc) reads on a timeout. Since there's an arbitrary amount of time between builds, no length of timeout is going to be sufficient.

It could be that once the client times out (and uses a much longer timeout...), it attempts to reconnect. I'm not sure what the consequences there would be, apart from constant re-opens of sockets. That might well be perfectly benign; I'm not sure.

Reply:
`Timeoute::Infinite` should do what you want, together with `read()` on a `ReadWriter`.

Reply:
I'll give it a whirl. If it's appropriate here, why not for ping, though?